rts/motoko-rts/src/types.rs

```diff
@@ -447,29 +457,91 @@ pub type Tag = u32;
 // locations in mark-compact GC. (Reminder: objects and fields are word aligned)
 pub const TAG_OBJECT: Tag = 1;
 pub const TAG_OBJ_IND: Tag = 3;
-pub const TAG_ARRAY: Tag = 5;
-pub const TAG_BITS64: Tag = 7;
-pub const TAG_MUTBOX: Tag = 9;
-pub const TAG_CLOSURE: Tag = 11;
-pub const TAG_SOME: Tag = 13;
-pub const TAG_VARIANT: Tag = 15;
-pub const TAG_BLOB: Tag = 17;
-pub const TAG_FWD_PTR: Tag = 19; // Only used by the copying GC - not to be confused with forwarding pointer in the header used for incremental GC.
-pub const TAG_BITS32: Tag = 21;
-pub const TAG_BIGINT: Tag = 23;
-pub const TAG_CONCAT: Tag = 25;
-pub const TAG_REGION: Tag = 27;
-pub const TAG_NULL: Tag = 29;
-pub const TAG_ONE_WORD_FILLER: Tag = 31;
-pub const TAG_FREE_SPACE: Tag = 33;
+pub const TAG_ARRAY_I: Tag = 5; // Immutable Array ([T])
+pub const TAG_ARRAY_M: Tag = 7; // Mutable Array ([var T])
+pub const TAG_ARRAY_T: Tag = 9; // Non-nullary Tuple ((T,+))
+pub const TAG_ARRAY_S: Tag = 11; // Shared function pairing TAG_BLOB_A with TAG_BLOB_T (shared ... -> ...)
+pub const TAG_BITS64_U: Tag = 13; // Unsigned (Nat64)
+pub const TAG_BITS64_S: Tag = 15; // Signed (Int64)
+pub const TAG_BITS64_F: Tag = 17; // Float
+pub const TAG_MUTBOX: Tag = 19;
+pub const TAG_CLOSURE: Tag = 21;
+pub const TAG_SOME: Tag = 23;
+pub const TAG_VARIANT: Tag = 25;
+pub const TAG_BLOB_B: Tag = 27; // Blob of Bytes (Blob)
+pub const TAG_BLOB_T: Tag = 29; // Blob of Utf8 (Text)
+pub const TAG_BLOB_P: Tag = 31; // Principal (Principal)
+pub const TAG_BLOB_A: Tag = 33; // Actor (actor {})
+pub const TAG_FWD_PTR: Tag = 35; // Only used by the copying GC - not to be confused with forwarding pointer in the header used for incremental GC.
+pub const TAG_BITS32_U: Tag = 37; // Unsigned (Nat32)
+pub const TAG_BITS32_S: Tag = 39; // Signed (Int32)
+pub const TAG_BITS32_F: Tag = 41; // Reserved (Float32)
+pub const TAG_BIGINT: Tag = 43;
+pub const TAG_CONCAT: Tag = 45;
+pub const TAG_REGION: Tag = 47;
+pub const TAG_NULL: Tag = 49;
+pub const TAG_ONE_WORD_FILLER: Tag = 51;
+pub const TAG_FREE_SPACE: Tag = 53;
 
 // Special value to visit only a range of array fields.
 // This and all values above it are reserved and mean
-// a slice of an array object (i.e. start index) for
+// a slice of an array object (i.e. compressed array tag + start index) for
 // purposes of `visit_pointer_fields`.
+// The top two bits encode the original array tag, the remaining bits are the start index of the slice.
 // Invariant: the value of this (pseudo-)tag must be
 //            higher than all other tags defined above
-pub const TAG_ARRAY_SLICE_MIN: Tag = 34;
+pub const TAG_ARRAY_SLICE_MIN: Tag = 54;
+
+pub const TAG_SPACING: Tag = 2;
+
+pub fn is_blob_tag(tag: Tag) -> bool {
+    tag == TAG_BLOB_B || tag == TAG_BLOB_T || tag == TAG_BLOB_P || tag == TAG_BLOB_A
```

Comment:
one could help the compiler doing such tests by arranging the tags to have values that only differ in the two penultimate bits, e.g. 29..35.

Reply:
True, but is_blob_tag is mostly called from assertions (apart from a relatively cheap call in crc32, that could also be an assert). Won't fix since Luc has already merged this PR into his stack...